### PR TITLE
fix: Debug buffer placement

### PR DESCRIPTION
**Pull request by joeyparrish**
Issue an error with debug info, including the URL, if the placement of
a segment in buffer is very different from expectations.

b/233075535

### CODE REVIEW

File: lib/media/time_ranges_utils.js [hunk 1]
```diff
@@ -155,4 +155,46 @@ shaka.media.TimeRangesUtils = class {
...
+    if (!oldRanges || !oldRanges.length) {
+      return null;
+    }
+    if (!newRanges || !newRanges.length) {
+      return TimeRangesUtils.getBufferedInfo(newRanges).pop();
+    }
+
+    const added = [];
+    for (const newRange of TimeRangesUtils.getBufferedInfo(newRanges)) {
+      let foundOverlap = false;
```
Comment (theodab) [inline lib/media/time_ranges_utils.js:179]: Perhaps I'm not thinking of some possible scenario here, but couldn't this algorithm potentially be a little faster for super-messy inputs if you read `newRanges` and `oldRanges` in reverse order, and returned if you saw a non-overlap or a newly-added difference? That would return early in some circumstances, and would remove the need to keep an array of added ranges. Something like...

```js
    const reversedNewRanges = TimeRangesUtils.getBufferedInfo(newRanges).reverse();
    const reversedOldRanges = TimeRangesUtils.getBufferedInfo(oldRanges).reverse();
    for (const newRange of reversedNewRanges) {
      let foundOverlap = false;
      for (const oldRange of reversedOldRanges) {
        if (oldRange.end >= newRange.start && oldRange.end <= newRange.end) {
          foundOverlap = true;

          // If the new range goes beyond the corresponding old one, the
          // difference is newly-added.
          if (newRange.end > oldRange.end) {
            return {start: oldRange.end, end: newRange.end};
          }
        }
      }

      if (!foundOverlap) {
        return newRange;
      }
    }

    return null;
```

**Review by theodab — CHANGES_REQUESTED** [review]

File: lib/media/media_source_engine.js [hunk 2]
```diff
@@ -569,26 +574,27 @@ shaka.media.MediaSourceEngine = class {
...
-              closedCaptions, startTime, endTime, videoOffset);
+              closedCaptions, reference.startTime, reference.endTime,
+              videoOffset);
         }
       }
     }
 
-    data = this.workAroundBrokenPlatforms_(data, startTime, contentType);
+    data = this.workAroundBrokenPlatforms_(
+        data, reference ? reference.startTime : null, contentType);
```
[the remark this answers is not in the document]
Comment (joeyparrish) [reply lib/media/media_source_engine.js:592]: I wrote those two things on different days, and with different amounts of whitespace available on the line.  :-)

I'll make it consistent.

File: lib/media/time_ranges_utils.js (hunk 1 above)
Comment (joeyparrish) [reply lib/media/time_ranges_utils.js:179]: You're right.  I started off trying to build something very general, then gave up and hacked it to work better for our purposes.  Your way is better.

File: lib/media/media_source_engine.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (joeyparrish) [reply lib/media/media_source_engine.js:592]: BTW, really looking forward to using "optional chaining": https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

[the remark this answers is not in the document]
Comment (joeyparrish) [reply lib/media/media_source_engine.js:592]: Though it looks like that will have to wait until we move to TypeScript some day.

File: lib/media/time_ranges_utils.js (hunk 1 above)
Comment (joeyparrish) [reply lib/media/time_ranges_utils.js:182]: Done

File: lib/media/media_source_engine.js [hunk 3]
```diff
@@ -533,7 +533,8 @@ shaka.media.MediaSourceEngine = class {
         const offset = await this.textSequenceModeOffset_;
         this.textEngine_.setTimestampOffset(offset);
       }
-      await this.textEngine_.appendBuffer(data, startTime, endTime);
+      await this.textEngine_.appendBuffer(
+          data, reference.startTime, reference.endTime);
```
[the remark this answers is not in the document]
Comment (joeyparrish) [reply lib/media/media_source_engine.js]: Done

File: lib/media/media_source_engine.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (joeyparrish) [reply lib/media/media_source_engine.js:598]: Done

[the remark this answers is not in the document]
Comment (theodab) [reply lib/media/media_source_engine.js:598]: Optional chaining would be nice, yeah. I really like the stuff that Swift does with optionals, for instance.

**Review by theodab — APPROVED** [review]

File: lib/media/time_ranges_utils.js [hunk 4]
```diff
@@ -155,4 +155,49 @@ shaka.media.TimeRangesUtils = class {
...
     return ret;
   }
+
+  /**
+   * This operation can be potentially EXPENSIVE and should only be done in
+   * debug builds for debugging purposes.
+   *
+   * @param {TimeRanges} oldRanges
+   * @param {TimeRanges} newRanges
+   * @return {?shaka.extern.BufferedRange}
```
Comment (theodab) [inline lib/media/time_ranges_utils.js:165]: Actually, wait, one last thing.
This should probably specify that this is the last added range, chronologically.

**Review by theodab — CHANGES_REQUESTED** [review]

**Pull request merged**